File: src/dotnet/PloyBus.Tests/Transport/Transactions/Messages/MessageInfoTests.cs
using NUnit.Framework;

// ReSharper disable SuspiciousTypeConversion.Global

namespace PolyBus.Transport.Transactions.Messages;

[TestFixture]
public class MessageInfoTests
{
    [Test]
    public void GetAttributeFromHeader_WithValidHeader_ReturnsCorrectAttribute()
    {
        // Arrange
        var header = "endpoint=user-service, type=Command, name=create-user, version=1.2.3";

        // Act
        var result = MessageInfo.GetAttributeFromHeader(header);

        // Assert
        Assert.That(result, Is.Not.Null);
        Assert.That(result.Endpoint, Is.EqualTo("user-service"));
        Assert.That(result.Type, Is.EqualTo(MessageType.Command));
        Assert.That(result.Name, Is.EqualTo("create-user"));
        Assert.That(result.Major, Is.EqualTo(1));
        Assert.That(result.Minor, Is.EqualTo(2));
        Assert.That(result.Patch, Is.EqualTo(3));
    }

    [Test]
    public void GetAttributeFromHeader_WithEventType_ReturnsCorrectAttribute()
    {
        // Arrange
        var header = "endpoint=notification-service, type=Event, name=user-created, version=2.0.1";

        // Act
        var result = MessageInfo.GetAttributeFromHeader(header);

        // Assert
        Assert.That(result, Is.Not.Null);
        Assert.That(result!.Endpoint, Is.EqualTo("notification-service"));
        Assert.That(result.Type, Is.EqualTo(MessageType.Event));
        Assert.That(result.Name, Is.EqualTo("user-created"));
        Assert.That(result.Major, Is.EqualTo(2));
        Assert.That(result.Minor, Is.EqualTo(0));
        Assert.That(result.Patch, Is.EqualTo(1));
    }

    [Test]
    public void GetAttributeFromHeader_WithExtraSpaces_ReturnsCorrectAttribute()
    {
        // Arrange - the current regex doesn't handle spaces within values well, so testing valid spacing
        var header = "endpoint=payment-service, type=Command, name=process-payment, version=3.14.159";

        // Act
        var result = MessageInfo.GetAttributeFromHeader(header);

        // Assert
        Assert.That(result, Is.Not.Null);
        Assert.That(result!.Endpoint, Is.EqualTo("payment-service"));
        Assert.That(result.Type, Is.EqualTo(MessageType.Command));
        Assert.That(result.Name, Is.EqualTo("process-payment"));
        Assert.That(result.Major, Is.EqualTo(3));
        Assert.That(result.Minor, Is.EqualTo(14));
        Assert.That(result.Patch, Is.EqualTo(159));
    }

    [Test]
    public void GetAttributeFromHeader_WithCaseInsensitiveType_ReturnsCorrectAttribute()
    {
        // Arrange
        var header = "endpoint=order-service, type=command, name=place-order, version=1.0.0";

        // Act
        var result = MessageInfo.GetAttributeFromHeader(header);

        // Assert
        Assert.That(result, Is.Not.Null);
        Assert.That(result!.Type, Is.EqualTo(MessageType.Command));
    }

    [TestCase("")]
    [TestCase("invalid header")]
    [TestCase("endpoint=test")]
    [TestCase("endpoint=test, type=Command")]
    [TestCase("endpoint=test, type=Command, name=test")]
    [TestCase("endpoint=test, type=Command, name=test, version=invalid")]
    [TestCase("type=Command, name=Test, version=1.0.0")]
    public void GetAttributeFromHeader_WithInvalidHeader_ReturnsNull(string header)
    {
        // Act
        var result = MessageInfo.GetAttributeFromHeader(header);

        // Assert
        Assert.That(result, Is.Null);
    }

    [Test]
    public void GetAttributeFromHeader_WithInvalidEnumType_ThrowsArgumentException()
    {
        // Arrange
        var header = "endpoint=test, type=InvalidType, name=test, version=1.0.0";

        // Act & Assert
        Assert.Throws<ArgumentException>(() => MessageInfo.GetAttributeFromHeader(header));
    }

    [Test]
    public void GetAttributeFromHeader_WithMissingVersion_ReturnsNull()
    {
        // Arrange
        var header = "endpoint=test-service, type=Command, name=test-command, version=";

        // Act
        var result = MessageInfo.GetAttributeFromHeader(header);

        // Assert
        Assert.That(result, Is.Null);
    }

    [Test]
    public void GetAttributeFromHeader_WithIncompleteVersion_ReturnsNull()
    {
        // Arrange
        var header = "endpoint=test-service, type=Command, name=test-command, version=1.0";

        // Act
        var result = MessageInfo.GetAttributeFromHeader(header);

        // Assert
        Assert.That(result, Is.Null);
    }

    [Test]
    public void Equals_WithIdenticalAttributes_ReturnsTrue()
    {
        // Arrange
        var attr1 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var attr2 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);

        // Act & Assert
        Assert.That(attr1.Equals(attr2), Is.True);
        Assert.That(attr2.Equals(attr1), Is.True);
    }

    [Test]
    public void Equals_WithSameObject_ReturnsTrue()
    {
        // Arrange
        var attr = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);

        // Act & Assert
        Assert.That(attr.Equals(attr), Is.True);
    }

    [Test]
    public void Equals_WithDifferentType_ReturnsFalse()
    {
        // Arrange
        var attr1 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var attr2 = new MessageInfo(MessageType.Event, "user-service", "create-user", 1, 2, 3);

        // Act & Assert
        Assert.That(attr1.Equals(attr2), Is.False);
    }

    [Test]
    public void Equals_WithDifferentEndpoint_ReturnsFalse()
    {
        // Arrange
        var attr1 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var attr2 = new MessageInfo(MessageType.Command, "order-service", "create-user", 1, 2, 3);

        // Act & Assert
        Assert.That(attr1.Equals(attr2), Is.False);
    }

    [Test]
    public void Equals_WithDifferentName_ReturnsFalse()
    {
        // Arrange
        var attr1 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var attr2 = new MessageInfo(MessageType.Command, "user-service", "update-user", 1, 2, 3);

        // Act & Assert
        Assert.That(attr1.Equals(attr2), Is.False);
    }

    [Test]
    public void Equals_WithDifferentMajorVersion_ReturnsFalse()
    {
        // Arrange
        var attr1 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var attr2 = new MessageInfo(MessageType.Command, "user-service", "create-user", 2, 2, 3);

        // Act & Assert
        Assert.That(attr1.Equals(attr2), Is.False);
    }

    [Test]
    public void Equals_WithDifferentMinorVersion_ReturnsTrue()
    {
        // Arrange
        var attr1 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var attr2 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 3, 3);

        // Act & Assert
        Assert.That(attr1.Equals(attr2), Is.True, "Minor version differences should not affect equality");
    }

    [Test]
    public void Equals_WithDifferentPatchVersion_ReturnsTrue()
    {
        // Arrange
        var attr1 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var attr2 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 4);

        // Act & Assert
        Assert.That(attr1.Equals(attr2), Is.True, "Patch version differences should not affect equality");
    }

    [Test]
    public void Equals_WithNullObject_ReturnsFalse()
    {
        // Arrange
        var attr = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);

        // Act & Assert
        Assert.That(attr.Equals(null), Is.False);
    }

    [Test]
    public void Equals_WithDifferentObjectType_ReturnsFalse()
    {
        // Arrange
        var attr = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var other = "not a MessageAttribute";

        // Act & Assert
        Assert.That(attr.Equals(other), Is.False);
    }

    [Test]
    public void GetHashCode_WithIdenticalAttributes_ReturnsSameHashCode()
    {
        // Arrange
        var attr1 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var attr2 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);

        // Act & Assert
        Assert.That(attr1.GetHashCode(), Is.EqualTo(attr2.GetHashCode()));
    }

    [Test]
    public void GetHashCode_WithDifferentAttributes_ReturnsDifferentHashCodes()
    {
        // Arrange
        var attr1 = new MessageInfo(MessageType.Command, "user-service", "create-user", 1, 2, 3);
        var attr2 = new MessageInfo(MessageType.Event, "user-service", "create-user", 1, 2, 3);

        // Act & Assert
        Assert.That(attr1.GetHashCode(), Is.Not.EqualTo(attr2.GetHashCode()));
    }
}

File: src/kotlin/src/test/kotlin/polybus/transport/transactions/messages/MessageInfoTests.kt
package polybus.transport.transactions.messages

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

class MessageInfoTests {
    @Test
    fun getAttributeFromHeaderWithValidHeaderReturnsCorrectAttribute() {
        // Arrange
        val header = "endpoint=user-service, type=COMMAND, name=create-user, version=1.2.3"

        // Act
        val result = MessageInfoValue.getAttributeFromHeader(header)

        // Assert
        assertNotNull(result)
        assertEquals("user-service", result!!.endpoint)
        assertEquals(MessageType.COMMAND, result.type)
        assertEquals("create-user", result.name)
        assertEquals(1, result.major)
        assertEquals(2, result.minor)
        assertEquals(3, result.patch)
    }

    @Test
    fun getAttributeFromHeaderWithEventTypeReturnsCorrectAttribute() {
        // Arrange
        val header = "endpoint=notification-service, type=EVENT, name=user-created, version=2.0.1"

        // Act
        val result = MessageInfoValue.getAttributeFromHeader(header)

        // Assert
        assertNotNull(result)
        assertEquals("notification-service", result!!.endpoint)
        assertEquals(MessageType.EVENT, result.type)
        assertEquals("user-created", result.name)
        assertEquals(2, result.major)
        assertEquals(0, result.minor)
        assertEquals(1, result.patch)
    }

    @Test
    fun getAttributeFromHeaderWithExtraSpacesReturnsCorrectAttribute() {
        // Arrange - the current regex doesn't handle spaces within values well, so testing valid spacing
        val header = "endpoint=payment-service, type=COMMAND, name=process-payment, version=3.14.159"

        // Act
        val result = MessageInfoValue.getAttributeFromHeader(header)

        // Assert
        assertNotNull(result)
        assertEquals("payment-service", result!!.endpoint)
        assertEquals(MessageType.COMMAND, result.type)
        assertEquals("process-payment", result.name)
        assertEquals(3, result.major)
        assertEquals(14, result.minor)
        assertEquals(159, result.patch)
    }

    @Test
    fun getAttributeFromHeaderWithCaseInsensitiveTypeReturnsCorrectAttribute() {
        // Arrange
        val header = "endpoint=order-service, type=command, name=place-order, version=1.0.0"

        // Act
        val result = MessageInfoValue.getAttributeFromHeader(header)

        // Assert
        assertNotNull(result)
        assertEquals(MessageType.COMMAND, result!!.type)
    }

    @ParameterizedTest
    @ValueSource(
        strings = [
            "",
            "invalid header",
            "endpoint=test",
            "endpoint=test, type=COMMAND",
            "endpoint=test, type=COMMAND, name=test",
            "endpoint=test, type=COMMAND, name=test, version=invalid",
            "type=COMMAND, name=test, version=1.0.0"
        ]
    )
    fun getAttributeFromHeaderWithInvalidHeaderReturnsNull(header: String) {
        // Act
        val result = MessageInfoValue.getAttributeFromHeader(header)

        // Assert
        assertNull(result)
    }

    @Test
    fun getAttributeFromHeaderWithInvalidEnumTypeThrowsIllegalArgumentException() {
        // Arrange
        val header = "endpoint=test, type=INVALID_TYPE, name=test, version=1.0.0"

        // Act & Assert
        assertThrows(IllegalArgumentException::class.java) {
            MessageInfoValue.getAttributeFromHeader(header)
        }
    }

    @Test
    fun getAttributeFromHeaderWithMissingVersionReturnsNull() {
        // Arrange
        val header = "endpoint=test-service, type=COMMAND, name=test-command, version="

        // Act
        val result = MessageInfoValue.getAttributeFromHeader(header)

        // Assert
        assertNull(result)
    }

    @Test
    fun getAttributeFromHeaderWithIncompleteVersionReturnsNull() {
        // Arrange
        val header = "endpoint=test-service, type=COMMAND, name=test-command, version=1.0"

        // Act
        val result = MessageInfoValue.getAttributeFromHeader(header)

        // Assert
        assertNull(result)
    }

    @Test
    fun equalsWithIdenticalAttributesReturnsTrue() {
        // Arrange
        val attr1 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)
        val attr2 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)

        // Act & Assert
        assertTrue(attr1.equals(attr2))
        assertTrue(attr2.equals(attr1))
    }

    @Test
    fun equalsWithSameObjectReturnsTrue() {
        // Arrange
        val attr = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)

        // Act & Assert
        assertTrue(attr.equals(attr))
    }

    @Test
    fun equalsWithDifferentTypeReturnsFalse() {
        // Arrange
        val attr1 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)
        val attr2 = MessageInfoValue(MessageType.EVENT, "user-service", "create-user", 1, 2, 3)

        // Act & Assert
        assertFalse(attr1.equals(attr2))
    }

    @Test
    fun equalsWithDifferentMajorVersionReturnsFalse() {
        // Arrange
        val attr1 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)
        val attr2 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 2, 2, 3)

        // Act & Assert
        assertFalse(attr1.equals(attr2))
    }

    @Test
    fun equalsWithDifferentMinorVersionReturnsTrue() {
        // Arrange
        val attr1 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)
        val attr2 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 3, 3)

        // Act & Assert
        assertTrue(attr1.equals(attr2))
    }

    @Test
    fun equalsWithDifferentPatchVersionReturnsTrue() {
        val attr1 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)
        val attr2 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 4)

        // Act & Assert
        assertTrue(attr1.equals(attr2))
    }

    @Test
    fun equalsWithNullObjectReturnsFalse() {
        // Arrange
        val attr = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)

        // Act & Assert
        assertFalse(attr.equals(null))
    }

    @Test
    fun equalsWithDifferentObjectTypeReturnsFalse() {
        // Arrange
        val attr = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)
        val other: Any = "not a MessageAttribute"

        // Act & Assert
        assertFalse(attr.equals(other))
    }

    @Test
    fun getHashCodeWithIdenticalAttributesReturnsSameHashCode() {
        // Arrange
        val attr1 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)
        val attr2 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)

        // Act & Assert
        assertEquals(attr1.hashCode(), attr2.hashCode())
    }

    @Test
    fun getHashCodeWithDifferentAttributesReturnsDifferentHashCodes() {
        // Arrange
        val attr1 = MessageInfoValue(MessageType.COMMAND, "user-service", "create-user", 1, 2, 3)
        val attr2 = MessageInfoValue(MessageType.EVENT, "user-service", "create-user", 1, 2, 3)

        // Act & Assert
        assertNotEquals(attr1.hashCode(), attr2.hashCode())
    }

}
